lib/espresso-runner.js

```diff
@@ -3,38 +3,47 @@ import { JWProxy } from 'appium-base-driver';
 import { retryInterval } from 'asyncbox';
 import logger from './logger';
 import path from 'path';
-import { fs } from 'appium-support';
+import { fs, util } from 'appium-support';
+import { version } from '../../package.json'; // eslint-disable-line import/no-unresolved
 
 
 const TEST_APK_PATH = path.resolve(__dirname, '..', '..', 'espresso-server', 'app', 'build', 'outputs', 'apk', 'androidTest', 'debug', 'app-debug-androidTest.apk');
 const TEST_MANIFEST_PATH = path.resolve(__dirname, '..', '..', 'espresso-server', 'AndroidManifest-test.xml');
 const TEST_APK_PKG = "io.appium.espressoserver.test";
-const REQUIRED_PARAMS = ['adb', 'tmpDir', 'host', 'systemPort', 'devicePort', 'appPackage'];
+const REQUIRED_PARAMS = ['adb', 'tmpDir', 'host', 'systemPort', 'devicePort', 'appPackage', 'forceEspressoRebuild'];
 
 class EspressoRunner {
   constructor (opts = {}) {
     for (let req of REQUIRED_PARAMS) {
-      if (!opts || !opts[req]) {
+      if (!opts || !util.hasValue(opts[req])) {
         throw new Error(`Option '${req}' is required!`);
       }
       this[req] = opts[req];
     }
     this.jwproxy = new JWProxy({host: this.host, port: this.systemPort});
     this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
 
-    this.modServerPath = path.resolve(this.tmpDir, `${TEST_APK_PKG}.apk`);
+    this.modServerPath = path.resolve(this.tmpDir, `${TEST_APK_PKG}_${version}_${this.appPackage}.apk`);
   }
 
-  // TODO mostly duplicated from uiautomator2's installServerApk
   async installTestApk () {
-    if (await this.adb.isAppInstalled(TEST_APK_PKG)) {
-      // for now, uninstall always
-      // TODO: figure out how to know if the server is able to instrument the AUT
-      await this.adb.uninstallApk(TEST_APK_PKG);
+    if (this.forceEspressoRebuild) {
+      logger.debug(`Capability 'forceEspressoRebuild' on. Deleting file '${this.modServerPath}'`);
+      await fs.unlink(this.modServerPath);
     }
 
-    let tmpTestApkPath = await this.buildNewModServer();
-    await this.signAndInstall(this.modServerPath, TEST_APK_PKG);
+    let needsUninstall = false;
+    if (!(await fs.exists(this.modServerPath))) {
```

Comment:
Extract what? 

The value of `this.modServerPath` is of the form `/path/to/io.appium.espressoserver.test_<version of driver>_<package under test>.apk`.

So for our ApiDemos, it is `/path/to/io.appium.espressoserver.test_0.0.1_io.appium.android.apis.apk`. When we upgrade, say to `appium-espresso-driver` `0.0.2`, it will become `/path/to/io.appium.espressoserver.test_0.0.2_io.appium.android.apis.apk`, which will not be found, and trigger a rebuild with the latest server code.